File: Packages/OmniaApplication/Sources/OmniaApplication/ConversationDraftService.swift
import OmniaDomain

/// Durable unsent text drafts keyed by conversation identity. Configuration is
/// the single persistence source for drafts; conversation history remains the
/// source for accepted messages.
public actor ConversationDraftService {
    private let configurationService: ConfigurationService

    public init(configurationService: ConfigurationService) {
        self.configurationService = configurationService
    }

    public func draft(for conversation: ConversationIdentity) async throws -> String {
        try await configurationService.value(
            for: Self.key(for: conversation),
            at: .workspaceOverride
        ) ?? ""
    }

    public func save(
        _ draft: String,
        for conversation: ConversationIdentity
    ) async throws {
        if draft.isEmpty {
            try await remove(for: conversation)
        } else {
            try await configurationService.store(
                draft,
                for: Self.key(for: conversation),
                at: .workspaceOverride
            )
        }
    }

    public func remove(for conversation: ConversationIdentity) async throws {
        try await configurationService.remove(
            Self.key(for: conversation),
            at: .workspaceOverride
        )
    }

    public static func key(
        for conversation: ConversationIdentity
    ) -> ConfigurationKey<String> {
        ConfigurationKey("conversationDraft.\(conversation.canonicalString)")
    }
}

File: Android/core/application/src/main/kotlin/com/omnia/application/ConversationDraftService.kt
package com.omnia.application

import com.omnia.domain.ConfigurationKey
import com.omnia.domain.ConversationIdentity

class ConversationDraftService(
    private val configurationService: ConfigurationService,
) {
    suspend fun draft(conversation: ConversationIdentity): String =
        configurationService.resolved(key(conversation)) ?: ""

    suspend fun save(text: String, conversation: ConversationIdentity) {
        if (text.isBlank()) {
            configurationService.remove(key(conversation), DRAFT_LEVEL)
        } else {
            configurationService.store(text, key(conversation), DRAFT_LEVEL)
        }
    }

    suspend fun remove(conversation: ConversationIdentity) {
        configurationService.remove(key(conversation), DRAFT_LEVEL)
    }

    companion object {
        private val DRAFT_LEVEL = com.omnia.domain.ConfigurationLevel.workspaceOverride

        fun key(conversation: ConversationIdentity): ConfigurationKey<String> =
            ConfigurationKey("conversationDraft.${conversation.id}")
    }
}
